File: ios/WebViewController.m
//
//  WebViewController.m
//  Flyy
//
//  Created by Pooja Deshpande on 08/01/21.
//  Copyright © 2021 Facebook. All rights reserved.
//

#import "WebViewController.h"
#import <WebKit/WebKit.h>
#import "FlyyFramework/FlyyFramework.h"

@interface WebViewController ()<WKUIDelegate,WKNavigationDelegate,WKScriptMessageHandler>

@property (strong,nonatomic) WKWebView *wkWebView;
@property (strong,nonatomic) UILabel *loadingOffer;

@end

@implementation WebViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    //    self.title = @"Offer screen";
    self.view.backgroundColor = [UIColor whiteColor];
    self.navigationController.navigationBarHidden = YES;
    
    // Add and load URL in Webview
    
    WKWebViewConfiguration *configuration = [[WKWebViewConfiguration alloc] init];
    self.wkWebView = [[WKWebView alloc] initWithFrame:self.view.frame configuration:configuration];
    self.wkWebView.navigationDelegate = self;
    
    NSURL *nsurl=[NSURL URLWithString:self.pageUrl];
    NSURLRequest *nsrequest=[NSURLRequest requestWithURL:nsurl];
    [self.wkWebView loadRequest:nsrequest];
    
    NSString *scriptSource = [NSString stringWithFormat:@"%@ %@ %s %@ %@",@"var meta = document.createElement('meta');",@"meta.name = 'viewport';","meta.content = 'width=device-width, initial-scale=1.0, maximum- scale=1.0, user-scalable=no';",@"var head = document.getElementsByTagName('head')[0];",@"head.appendChild(meta);"];
    WKUserScript *userScript = [[WKUserScript alloc] initWithSource:scriptSource
                                                      injectionTime:WKUserScriptInjectionTimeAtDocumentEnd
                                                   forMainFrameOnly:YES];
    [self.wkWebView.configuration.userContentController addUserScript:userScript];
    [self.wkWebView.configuration.userContentController addScriptMessageHandler:self name:@"handleMessage"];
    [self.view addSubview:self.wkWebView];
    
    // Add label to show message
    
    self.loadingOffer = [[UILabel alloc]initWithFrame:CGRectMake(0, 0, 200, 30)];
    self.loadingOffer.text = self.pageLoadingTitle;
    self.loadingOffer.font = [UIFont fontWithName:@"Helvetica" size:18.0];
    self.loadingOffer.backgroundColor = [UIColor clearColor];
    self.loadingOffer.textColor = [UIColor blackColor];
    self.loadingOffer.textAlignment = NSTextAlignmentCenter;
    [self.view addSubview:self.loadingOffer];
    
}

-(void) viewDidLayoutSubviews
{
    self.loadingOffer.center = self.view.center;
}

-(void)webView:(WKWebView *)webView didFailNavigation:(WKNavigation *)navigation withError:(NSError *)error
{
    self.loadingOffer.text = @"";
    self.loadingOffer.hidden = YES;
}

-(void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation
{
    [self sendDataToJS];
    self.loadingOffer.text = @"";
    self.loadingOffer.hidden = YES;
}

- (void)userContentController:(WKUserContentController *)userContentController didReceiveScriptMessage:(WKScriptMessage *)message
{
    NSLog(@"userContentController:didReceiveScriptMessage: message=%@", message);
    if ([message.name isEqualToString:@"handleMessage"])
    {
        if ([message.body isKindOfClass:[NSDictionary class]])
        {
            NSDictionary *body = message.body;
            NSString *param = body[@"action"];
            if ([param isEqualToString:@"go-back"])
            {
                if(self.wkWebView.canGoBack)
                {
                    [self.wkWebView goBack];
                }
                else
                {
                    [self.navigationController popViewControllerAnimated:YES];
                }
            }
            else
            {
                return;
            }
        }
    }
}

-(void) sendDataToJS
{
    Flyy *flyyInstance = [[Flyy alloc] init];
    
    NSString *jsData = [NSString stringWithFormat:@"handleSDKMessage('%@')",[flyyInstance getHeadersDataForWebView]];
    [self.wkWebView evaluateJavaScript:jsData completionHandler:^(NSString *result, NSError * _Nullable error) {
        NSLog(@"%@",result);
    }];
}

@end

File: ios/WebViewController.swift
//
//  WebViewController.swift
//  react-native-flyy
//
//  Created by Pooja Deshpande on 06/01/21.
//

import Foundation
import UIKit
import WebKit
import FlyyFramework

internal class WebViewController: UIViewController, WKUIDelegate, WKNavigationDelegate, WKScriptMessageHandler {
    
    var webView:WKWebView!
    var label:UILabel!
    
    public static var PAGE_TO_OPEN = URL(string: "")
    public static var LABEL_TEXT = "Loading Offers..."
    
    override func loadView() {
        webView = WKWebView()
               webView.uiDelegate = self
               webView.navigationDelegate = self
               view = webView
        
        label = UILabel(frame: CGRect(x: 0, y: 0, width: 200, height: 30))
        label.numberOfLines = 0
        label.text = WebViewController.LABEL_TEXT
        
        label.textAlignment = .center
        label.font = UIFont(name: "Helvetica", size: 18.0)
        label.backgroundColor = UIColor.clear
        
        self.view.addSubview(label)
    }
    
    public override func viewDidLayoutSubviews() {
            label.center = self.view.center
        }
    
    public override func viewDidLoad() {
        super.viewDidLoad()
        
        let request = URLRequest(url: WebViewController.PAGE_TO_OPEN!)
        
        webView.load(request)
        webView.allowsBackForwardNavigationGestures = true
        webView.configuration.userContentController.addUserScript(self.getZoomDisableScript())
        webView.configuration.userContentController.add(self, name: "handleMessage")
    }
    
    public func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        label.text = ""
        label.isHidden = true
    }
    
    public func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        sendDataToJS()
        label.text = ""
        label.isHidden = true
    }
    
    public func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        if(error._code == NSURLErrorNotConnectedToInternet){
            showNoInternetConnection()
        }
    }
    
    func showNoInternetConnection() {
        // create the alert
        let alert = UIAlertController(title: "No Internet Connection", message: "Make sure your device is connected to the internet.", preferredStyle: UIAlertController.Style.alert)
        
        // add an action (button)
        alert.addAction(UIAlertAction(title: "OK", style: UIAlertAction.Style.default, handler: { (action: UIAlertAction!) in
            //            self.navigationController?.popViewController(animated: true)
            self.dismiss(animated: true, completion: nil)
        }))
        // show the alert
        self.present(alert, animated: true, completion: nil)
    }
    
    public func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        if (message.name == "handleMessage") {
            guard let body = message.body as? [String: AnyObject],
                  let param = body["action"] as? String else {
                return
            }
            
            if(param == "go-back") {
                if(webView.canGoBack) {
                    //Go back in webview history
                    webView.goBack()
                } else {
                    //Pop view controller to preview view controller
                    //                    self.navigationController?.popViewController(animated: true)
                    self.dismiss(animated: true, completion: nil)
                }
            }
        }
        
    }
    
    private func getZoomDisableScript() -> WKUserScript {
        let source: String = "var meta = document.createElement('meta');" +
            "meta.name = 'viewport';" +
            "meta.content = 'width=device-width, initial-scale=1.0, maximum- scale=1.0, user-scalable=no';" +
            "var head = document.getElementsByTagName('head')[0];" + "head.appendChild(meta);"
        return WKUserScript(source: source, injectionTime: .atDocumentEnd, forMainFrameOnly: true)
    }
    
    func sendDataToJS() {
        // Send the header data to the page
        let javascript = "handleSDKMessage('\(Flyy.sharedInstance.getHeadersDataForWebView())')"
        
        webView.evaluateJavaScript(javascript) { (result, error) in
            guard result == nil else {
                if let unwrapped = result {
                    debugPrint(unwrapped)
                }
                return
            }
            
            guard error == nil else {
                debugPrint("Inside Flyy SDK JS error \(error!)")
                return
            }
        }
    }
    
}
